### PR TITLE
uart: Use DMA for uart_tx

### LINKED ISSUE #21: uart: Bringup basic functionality
Remember to toggle the analog switches using the max14662 driver.

### PR DESCRIPTION
**Pull request by kbeckmann**
This commit prepares for uart_rx_dma as well.

Closes #21

### CODE REVIEW

File: app/config/FreeRTOSConfig.h [hunk 1]
```diff
@@ -115,6 +115,8 @@ extern uint32_t SystemCoreClock;
 #define configTIMER_QUEUE_LENGTH                 32
 #define configTIMER_TASK_STACK_DEPTH             512
 
+#define configUSE_COUNTING_SEMAPHORES 1
```
Comment (kbeckmann) [inline app/config/FreeRTOSConfig.h]: remove

File: app/drivers/adc.c [hunk 2]
```diff
@@ -36,7 +36,7 @@ err_t adc_init(void)
 	adc_handle.Init.DMAContinuousRequests = DISABLE;
 	adc_handle.Init.Overrun = ADC_OVR_DATA_PRESERVED;
 	adc_handle.Init.OversamplingMode = ENABLE;
-	adc_handle.Init.Oversampling.Ratio = ADC_OVERSAMPLING_RATIO_16;
+	adc_handle.Init.Oversampling.Ratio = ADC_OVERSAMPLING_RATIO_256;
```
Comment (kbeckmann) [inline app/drivers/adc.c]: remove

File: app/drivers/uart.h [hunk 3]
```diff
@@ -4,6 +4,11 @@
 #include "errors.h"
 
 #define EUART_HAL_INIT					(EUART_BASE + 0)
+#define EUART_TX_SEMPH					(EUART_BASE + 1)
+#define EUART_TX						(EUART_BASE + 2)
+#define EUART_RX						(EUART_BASE + 3)
 
 UART_HandleTypeDef *uart_get_handle(void);
 err_t uart_init(void);
+err_t uart_tx(UART_HandleTypeDef *p_handle, const uint8_t *p_buf, uint32_t size, uint32_t timeout_ms);
```
Comment (arturo182) [inline app/drivers/uart.h]: could you add them before init, so init is last? ;)

File: app/drivers/uart.h [hunk 4]
```diff
@@ -4,6 +4,11 @@
 #include "errors.h"
 
 #define EUART_HAL_INIT					(EUART_BASE + 0)
+#define EUART_TX_SEMPH					(EUART_BASE + 1)
+#define EUART_TX						(EUART_BASE + 2)
+#define EUART_RX						(EUART_BASE + 3)
 
 UART_HandleTypeDef *uart_get_handle(void);
+err_t uart_tx(UART_HandleTypeDef *p_handle, const uint8_t *p_buf, uint32_t size, uint32_t timeout_ms);
+err_t uart_rx(UART_HandleTypeDef *p_handle, uint8_t *p_buf, uint32_t size, uint32_t timeout_ms);
```
Comment (kbeckmann) [inline app/drivers/uart.h]: maybe we should use (freertos-)ticks instead for the timeout parameters.

Comment (arturo182) [reply app/drivers/uart.h]: UT the Hal uses ms so we would convert one way and then back to pass to the Hal?

**Pull request merged**